AK+Tests+Everywhere: Fix UTF-8-related crashes

BenWiederhake opened the pull request:
This PR:
- Fixes silent data corruption due to `StringView::contains(char)` weirdly accepting (and truncating) `u32` code points, which is one cause of the crash https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49184 .
- Fixes invalid code points like `U+123456` being produced by `Utf8CodePointIterator`, which is the other cause of the crash https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=49184 .
- Fixes a slightly-wrong function signature as a drive-by fix.

Due to this, this PR should close the OSS Fuzz issue 49184 for good, and prevent future such issues.

Also, all the calls to `StringView::contains(char)` are now guaranteed to be correct, because either the type matches exactly, or the compiler complains about ambiguous types (e.g. `u16` being passed in, which is why I had to touch LibJS and LibWeb).

ADKaster commented on Userland/Libraries/LibJS/Runtime/GlobalObject.cpp:
```diff
@@ -350,7 +350,7 @@ static ThrowCompletionOr<String> encode(VM& vm, String const& string, StringView
         // c. If C is in unescapedSet, then
         // NOTE: We assume the unescaped set only contains ascii characters as unescaped_set is a StringView.
-        if (code_unit < 0x80 && unescaped_set.contains(code_unit)) {
+        if (code_unit < 0x80 && unescaped_set.contains((char)code_unit)) {
```
Can we avoid c-style casts in new code? static_cast pls (global)

BenWiederhake replied:
Ah, right! I'll change the other casts I introduced.

BenWiederhake commented:
Changes since last push:
- Changed C-style casts to `static_cast`s (thanks @ADKaster!)

trflynn89 commented on AK/StringView.cpp:
```diff
@@ -137,6 +138,18 @@ bool StringView::contains(char needle) const
 
+bool StringView::contains(u32 needle) const
+{
+    StringBuilder needle_builder;
```
A method named `contains` performing an allocation would be very surprising to me.

Can we instead put an `AK::Array<char, 4>` on the stack here? Then use `AK::UnicodeUtils::code_point_to_utf8` to encode the code point into the array. You can then invoke the `contains(StringView)` overload on that array.

Something like (untested):
```c++
Array<char, 4> encoded;
size_t index = 0;

auto length = AK::UnicodeUtils::code_point_to_utf8(needle, [&](auto byte) {
    encoded[index++] = byte;
});

if (length == -1)
    return false;

return contains(StringView { encoded.data(), length });
```

BenWiederhake replied:
There is no allocation here. In fact, I implemented the local-char-array you describe completely, before I noticed that StringBuilder already has an inline-buffer.

bgianfo replied:
Should there be a comment explaining that then? As I think many folks would read this code and assume the same. Especially with a call to a `try_` function.

BenWiederhake replied:
Given that we now have `NoAllocationGuard`, how about we chuck that in there?

BenWiederhake replied:
NoAllocationGuard doesn't easily work, since for some reason it can't be used by the Kernel(?!) or DynamicLoader. So I added a comment, like you suggested.

BenWiederhake commented:
Changes since last push:
- Added a comment about not allocating
- Rebased for convenience

IdanHo commented on AK/Utf8View.cpp:
```diff
@@ -274,6 +274,9 @@ u32 Utf8CodePointIterator::operator*() const
     }
 
+    if (code_point_value_so_far > 0x10FFFF) {
+        return 0xFFFD;
```
No need for curlys with single line blocks

BenWiederhake replied:
Ah, right. Personally I like keeping the curlys always, but I see that's not how we do it here.

Come to think about it: The other parts of the code do `dbgln_if`, so it would be *even more* in keeping the style to add a `dbgln_if`-statement, which then also justifies the curlys, so that's what I'll do. What do you think?

IdanHo replied:
sure, sounds good

IdanHo reviewed (commented):
One nitpick

BenWiederhake commented:
Changes since last push:
- Added a `dbgln_if` statement in case of rejected code point
- Rebased for convenience
- Re-validated that this still fixes the fuzzer crash
